File: e183/java/Solution.java
public class Solution {
    /** 
     *@param L: Given n pieces of wood with length L[i]
     *@param k: An integer
     *return: The maximum length of the small pieces.
     */
    public int woodCut(int[] L, int k) {
        // write your code here
        long sum = 0L;
        int longest = 0;
        for (int wood:L) {
            sum += (long)wood;
            if (wood > longest) {
                longest = wood;
            }
        }
        if (sum < (long)k) { //length = 1
            return 0;
        }
        if (divide(L, longest) >= k) {
            return longest;
        }

        int ret = 1;
        int low = 1;
        int high = longest;
        while (high - low > 1) {
            int mid = low / 2 + high / 2 + (low % 2 + high % 2) / 2;
            if (divide(L, mid) < k) {
                high = mid;
            }
            else {
                low = mid;
            }
        }
        return low;
    }
    public int divide(int[] L, int length) {
        int k = 0;
        for (int wood:L) {
            k += wood / length;
        }
        return k;
    }
}

File: e183/python/solution.py
class Solution:
    """
    @param L: Given n pieces of wood with length L[i]
    @param k: An integer
    return: The maximum length of the small pieces.
    """
    def woodCut(self, L, k):
        # write your code here
        if sum(L) < k:
            return 0
        longest = max(L)
        if sum([x / longest for x in L]) >= k:
            return longest
        low = 1
        high = longest
        ret = 1
        while high - low > 1:
            mid = (low + high) / 2
            n = sum([x / mid for x in L])
            if n < k:
                high = mid
            else: #n >= k
                low = mid
                if mid > ret:
                    ret = mid
        return ret
